Brokerages/Kraken/KrakenSymbolMapper.cs

```diff
@@ -0,0 +1,292 @@
+namespace QuantConnect.Brokerages.Kraken
+{
+    /// <summary>
+    /// Provides the mapping between Lean symbols and Kraken symbols.
+    /// </summary>
+    public class KrakenSymbolMapper : ISymbolMapper
+    {
+        class KrakenAsset
+        {   // currency code
+            public string Code;   
+            public string CodeAlt;
+        }
+
+        List<KrakenAsset> KnownAssets = new List<KrakenAsset>();
+
+        List<string> KnownPairs = new List<string>();
+        List<DataType.AssetPair> KnownPairsWhole = new List<DataType.AssetPair>();
+
+        static List<string> Prefixes = new List<string>() { "", "Z", "X" };
+
+        /// <summary>
+        /// Updates all Kraken Symbols
+        /// </summary>
+        /// <param name="restApi">Existing KrakenRestApi object</param>
+        public void UpdateSymbols(KrakenRestApi restApi)
+        {
+            Logging.Log.Trace("UpdateSymbols(restApi)");
+
+            Dictionary<string, DataType.AssetInfo> assetInfoDict = restApi.GetAssetInfo();
+            Dictionary<string, DataType.AssetPair> assetPairDict = restApi.GetAssetPairs();
+
+            KnownAssets.Clear();
+
+            foreach (KeyValuePair<string, DataType.AssetInfo> KVPair in assetInfoDict)
+            {
+                if (KVPair.Key == "KFEE")
+                    continue;
+
+                KrakenAsset krakenAsset = new KrakenAsset() { Code = KVPair.Key, CodeAlt = KVPair.Value.Altname };
+
+                KnownAssets.Add(krakenAsset);
+
+                Logging.Log.Trace($"KrakenAsset {{ Code: { KVPair.Key }, CodeAlt: { KVPair.Value.Altname } }}");
+            }
+
+            KnownPairs.Clear();
+            KnownPairsWhole.Clear();
+
+            foreach (KeyValuePair<string, DataType.AssetPair> KVPair in assetPairDict)
+            {
+                // remove dark pool markets
+                if (KVPair.Key.Contains(".d"))
+                    continue;
+
+                // KrakenPair krakenPair = new KrakenPair() { Pair = KVPair.Key, PairAlt = KVPair.Value.Altname };
+                KnownPairs.Add(KVPair.Key);
+                KnownPairsWhole.Add(KVPair.Value);
+
+                Logging.Log.Trace($"KnownPair {KVPair.Key}");
+            }
+        }
+         
+        void DecomposeKrakenPair(string krakenPair, out string baseCode, out string quoteCode)
+        {
+            Logging.Log.Trace($"DecomposeKrakenPair({krakenPair}, out baseCode, out quoteCode)");
+
+            List<string> foundCodes = new List<string>();
+
+            foreach(var krakenCode in KnownAssets)
+            {
+                if(krakenPair.Contains(krakenCode.Code))
+                {
+                    foundCodes.Add(krakenCode.Code);
+                }
+
+                if (krakenCode.Code != krakenCode.CodeAlt && krakenPair.Contains(krakenCode.CodeAlt))
+                {
+                    foundCodes.Add(krakenCode.CodeAlt);
+                }
+            }
+
+            foreach(string b in foundCodes)
```

Comment:
Yes, I know, I already have plan to make generalized splitting function with additional input of known symbols/codes :).

Should I write that function in crypto-symbol-length branch?

Reply:
All this stuff is pretty sensitive and can end up impacting many areas of code, I'd prefer if each thing is handled separately. It could be a refactor on what's already there and submitted as a separate PR. A nice part about refactoring in a separate PR is the goal is that nothing changes, so if the test coverage is good then the refactor is good when all tests are still passing (and maybe some new ones that touch on some interesting points in the code being refactored)